ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,138 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
 
 namespace
 {
-    Shape evaluate_matmul_output_shape(const Shape& arg0_shape,
-                                       const Shape& arg1_shape,
-                                       bool transpose_a,
-                                       bool transpose_b)
+    PartialShape validate_matmul_output_shape(const PartialShape& arg0_shape,
+                                              const PartialShape& arg1_shape,
+                                              bool transpose_a,
+                                              bool transpose_b)
     {
-        Shape output_shape;
-        Shape arg0_shape_update = arg0_shape;
-        Shape arg1_shape_update = arg1_shape;
+        auto arg0_rank = arg0_shape.rank().get_length();
+        auto arg1_rank = arg1_shape.rank().get_length();
+        auto max_rank = std::max(arg0_rank, arg1_rank);
+        auto merged_dimension = Dimension::dynamic();
 
-        size_t arg0_rank = arg0_shape.size();
-        size_t arg1_rank = arg1_shape.size();
+        std::vector<Dimension> arg0_shape_update(arg0_rank);
+        std::vector<Dimension> arg1_shape_update(arg1_rank);
+        std::vector<Dimension> output_shape(max_rank);
 
-        if (transpose_a && arg0_rank > 1)
+        for (size_t i = 0; i < arg0_rank; i++)
+            arg0_shape_update[i] = arg0_shape[i];
+        for (size_t i = 0; i < arg1_rank; i++)
+            arg1_shape_update[i] = arg1_shape[i];
+
+        if (transpose_a)
         {
+            // 1D vector of shape {S} is reshaped to {S, 1}
+            if (arg0_rank == 1)
+            {
+                arg0_shape_update.insert(arg0_shape_update.begin(), 1);
+                arg0_rank = arg0_shape_update.size();
+            }
+
             swap(arg0_shape_update[arg0_rank - 2], arg0_shape_update[arg0_rank - 1]);
         }
-        if (transpose_b && arg1_rank > 1)
+
+        if (transpose_b)
         {
+            // 1D vector of shape {S} is reshaped to {1, S}
+            if (arg1_rank == 1)
+            {
+                arg1_shape_update.insert(arg1_shape_update.begin(), 1);
+                arg1_rank = arg1_shape_update.size();
+            }
+
             swap(arg1_shape_update[arg1_rank - 2], arg1_shape_update[arg1_rank - 1]);
         }
 
         if (arg0_rank == 1 && arg1_rank == 1)
```

Comment:
Here and below, can we avoid checking rank by unsqueezing each input shapes up to 2D+?

Reply:
Yes, please refer to https://github.com/openvinotoolkit/openvino/pull/2212/files/8f5fb74a08bc6f27d88a1d5c4fa6070309e5b7f6#r500244697